File: Pinball App 2/Pinball App 2/gameroom/GameRoomStoreSnapshotSupport.swift
import Foundation

extension GameRoomStore {
    var activeMachines: [OwnedMachine] {
        state.ownedMachines
            .filter { $0.status.countsAsActiveInventory }
            .sorted(by: sortMachines)
    }

    var archivedMachines: [OwnedMachine] {
        state.ownedMachines
            .filter { !$0.status.countsAsActiveInventory }
            .sorted(by: sortMachines)
    }

    func snapshot(for machineID: UUID) -> OwnedMachineSnapshot {
        snapshots[machineID] ?? OwnedMachineSnapshot(
            ownedMachineID: machineID,
            currentPlayCount: 0,
            lastGlassCleanedAt: nil,
            lastPlayfieldCleanedAt: nil,
            lastPlayfieldCleanerUsed: nil,
            lastBallsServicedAt: nil,
            lastBallsReplacedAt: nil,
            currentBallSetNotes: nil,
            lastPitchCheckedAt: nil,
            currentPitchValue: nil,
            currentPitchMeasurementPoint: nil,
            lastLeveledAt: nil,
            lastRubberServiceAt: nil,
            lastFlipperServiceAt: nil,
            lastGeneralInspectionAt: nil,
            lastServiceAt: nil,
            openIssueCount: 0,
            dueTaskCount: 0,
            attentionState: .gray,
            updatedAt: Date()
        )
    }

    func recomputeSnapshots() {
        var nextSnapshots: [UUID: OwnedMachineSnapshot] = [:]
        let now = Date()

        for machine in state.ownedMachines {
            let machineEvents = gameRoomSortedMachineEventsForSnapshots(
                allEvents: state.events,
                machineID: machine.id
            )
            let machineIssues = gameRoomMachineIssues(
                allIssues: state.issues,
                machineID: machine.id
            )
            let openIssueCount = gameRoomOpenIssueCount(machineIssues)
            let currentPlayCount = gameRoomCurrentPlayCount(for: machineEvents)
            let dueTaskCount = gameRoomDueTaskCount(
                for: machine,
                events: machineEvents,
                reminderConfigs: state.reminderConfigs,
                currentPlayCount: currentPlayCount,
                now: now
            )
            let attentionState: GameRoomAttentionState = {
                if machine.status == .archived || machine.status == .sold || machine.status == .traded {
                    return .gray
                }
                if machineIssues.contains(where: { $0.status != .resolved && ($0.severity == .high || $0.severity == .critical) }) {
                    return .red
                }
                if openIssueCount > 0 || dueTaskCount > 0 {
                    return .yellow
                }
                return .green
            }()

            let latestEvent = gameRoomLatestEventLookup(events: machineEvents)

            let latestPitchCheck = latestEvent(.pitchChecked)

            nextSnapshots[machine.id] = OwnedMachineSnapshot(
                ownedMachineID: machine.id,
                currentPlayCount: currentPlayCount,
                lastGlassCleanedAt: latestEvent(.glassCleaned)?.occurredAt,
                lastPlayfieldCleanedAt: latestEvent(.playfieldCleaned)?.occurredAt,
                lastPlayfieldCleanerUsed: latestEvent(.playfieldCleaned)?.consumablesUsed,
                lastBallsServicedAt: latestEvent(.ballsCleaned)?.occurredAt,
                lastBallsReplacedAt: latestEvent(.ballsReplaced)?.occurredAt,
                currentBallSetNotes: latestEvent(.ballsReplaced)?.notes,
                lastPitchCheckedAt: latestPitchCheck?.occurredAt,
                currentPitchValue: latestPitchCheck?.pitchValue,
                currentPitchMeasurementPoint: latestPitchCheck?.pitchMeasurementPoint,
                lastLeveledAt: latestEvent(.machineLeveled)?.occurredAt,
                lastRubberServiceAt: latestEvent(.rubbersReplaced)?.occurredAt,
                lastFlipperServiceAt: latestEvent(.flipperServiced)?.occurredAt,
                lastGeneralInspectionAt: latestEvent(.generalInspection)?.occurredAt,
                lastServiceAt: machineEvents.first(where: { $0.category == .service })?.occurredAt,
                openIssueCount: openIssueCount,
                dueTaskCount: dueTaskCount,
                attentionState: attentionState,
                updatedAt: now
            )
        }

        snapshots = nextSnapshots
    }

    func sortMachines(lhs: OwnedMachine, rhs: OwnedMachine) -> Bool {
        let lhsArea = area(for: lhs.gameRoomAreaID)
        let rhsArea = area(for: rhs.gameRoomAreaID)

        let lhsAreaOrder = lhsArea?.areaOrder ?? Int.max
        let rhsAreaOrder = rhsArea?.areaOrder ?? Int.max
        if lhsAreaOrder != rhsAreaOrder { return lhsAreaOrder < rhsAreaOrder }

        let lhsAreaName = lhsArea?.name.lowercased() ?? ""
        let rhsAreaName = rhsArea?.name.lowercased() ?? ""
        if lhsAreaName != rhsAreaName { return lhsAreaName < rhsAreaName }

        let lhsGroup = lhs.groupNumber ?? Int.max
        let rhsGroup = rhs.groupNumber ?? Int.max
        if lhsGroup != rhsGroup { return lhsGroup < rhsGroup }

        let lhsPosition = lhs.position ?? Int.max
        let rhsPosition = rhs.position ?? Int.max
        if lhsPosition != rhsPosition { return lhsPosition < rhsPosition }

        let lhsTitle = lhs.displayTitle.lowercased()
        let rhsTitle = rhs.displayTitle.lowercased()
        if lhsTitle != rhsTitle { return lhsTitle < rhsTitle }

        return lhs.id.uuidString < rhs.id.uuidString
    }
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/gameroom/GameRoomStoreSnapshotSupport.kt
package com.pillyliu.pinprofandroid.gameroom

internal fun compareGameRoomMachines(
    lhs: OwnedMachine,
    rhs: OwnedMachine,
    areasById: Map<String, GameRoomArea>,
): Int {
    val lhsArea = lhs.gameRoomAreaID?.let(areasById::get)
    val rhsArea = rhs.gameRoomAreaID?.let(areasById::get)
    val lhsAreaOrder = lhsArea?.areaOrder ?: Int.MAX_VALUE
    val rhsAreaOrder = rhsArea?.areaOrder ?: Int.MAX_VALUE
    if (lhsAreaOrder != rhsAreaOrder) return lhsAreaOrder.compareTo(rhsAreaOrder)

    val lhsAreaName = lhsArea?.name?.lowercase().orEmpty()
    val rhsAreaName = rhsArea?.name?.lowercase().orEmpty()
    if (lhsAreaName != rhsAreaName) return lhsAreaName.compareTo(rhsAreaName)

    val lhsGroup = lhs.groupNumber ?: Int.MAX_VALUE
    val rhsGroup = rhs.groupNumber ?: Int.MAX_VALUE
    if (lhsGroup != rhsGroup) return lhsGroup.compareTo(rhsGroup)

    val lhsPosition = lhs.position ?: Int.MAX_VALUE
    val rhsPosition = rhs.position ?: Int.MAX_VALUE
    if (lhsPosition != rhsPosition) return lhsPosition.compareTo(rhsPosition)

    val lhsTitle = lhs.displayTitle.lowercase()
    val rhsTitle = rhs.displayTitle.lowercase()
    if (lhsTitle != rhsTitle) return lhsTitle.compareTo(rhsTitle)

    return lhs.id.compareTo(rhs.id)
}

internal fun computeOwnedMachineSnapshots(
    state: GameRoomPersistedState,
    nowMs: Long,
): Map<String, OwnedMachineSnapshot> {
    return state.ownedMachines.associate { machine ->
        val machineEvents = state.events
            .filter { it.ownedMachineID == machine.id }
            .sortedByDescending { it.occurredAtMs }
        val machineIssues = state.issues.filter { it.ownedMachineID == machine.id }
        val openIssueCount = machineIssues.count { it.status != MachineIssueStatus.resolved }
        val currentPlayCount = currentPlayCount(machineEvents)
        val dueTaskCount = dueTaskCount(
            machine = machine,
            events = machineEvents,
            currentPlayCount = currentPlayCount,
            nowMs = nowMs,
            reminderConfigs = state.reminderConfigs,
        )
        val attention = when {
            machine.status == OwnedMachineStatus.archived ||
                machine.status == OwnedMachineStatus.sold ||
                machine.status == OwnedMachineStatus.traded -> GameRoomAttentionState.gray

            machineIssues.any {
                it.status != MachineIssueStatus.resolved &&
                    (it.severity == MachineIssueSeverity.high || it.severity == MachineIssueSeverity.critical)
            } -> GameRoomAttentionState.red

            openIssueCount > 0 || dueTaskCount > 0 -> GameRoomAttentionState.yellow
            else -> GameRoomAttentionState.green
        }

        fun latestEvent(type: MachineEventType): MachineEvent? = machineEvents.firstOrNull { it.type == type }
        val latestPitch = latestEvent(MachineEventType.pitchChecked)

        machine.id to OwnedMachineSnapshot(
            ownedMachineID = machine.id,
            currentPlayCount = currentPlayCount,
            lastGlassCleanedAtMs = latestEvent(MachineEventType.glassCleaned)?.occurredAtMs,
            lastPlayfieldCleanedAtMs = latestEvent(MachineEventType.playfieldCleaned)?.occurredAtMs,
            lastPlayfieldCleanerUsed = latestEvent(MachineEventType.playfieldCleaned)?.consumablesUsed,
            lastBallsServicedAtMs = latestEvent(MachineEventType.ballsCleaned)?.occurredAtMs,
            lastBallsReplacedAtMs = latestEvent(MachineEventType.ballsReplaced)?.occurredAtMs,
            currentBallSetNotes = latestEvent(MachineEventType.ballsReplaced)?.notes,
            lastPitchCheckedAtMs = latestPitch?.occurredAtMs,
            currentPitchValue = latestPitch?.pitchValue,
            currentPitchMeasurementPoint = latestPitch?.pitchMeasurementPoint,
            lastLeveledAtMs = latestEvent(MachineEventType.machineLeveled)?.occurredAtMs,
            lastRubberServiceAtMs = latestEvent(MachineEventType.rubbersReplaced)?.occurredAtMs,
            lastFlipperServiceAtMs = latestEvent(MachineEventType.flipperServiced)?.occurredAtMs,
            lastGeneralInspectionAtMs = latestEvent(MachineEventType.generalInspection)?.occurredAtMs,
            lastServiceAtMs = machineEvents.firstOrNull { it.category == MachineEventCategory.service }?.occurredAtMs,
            openIssueCount = openIssueCount,
            dueTaskCount = dueTaskCount,
            attentionState = attention,
            updatedAtMs = nowMs,
        )
    }
}

private fun dueTaskCount(
    machine: OwnedMachine,
    events: List<MachineEvent>,
    currentPlayCount: Int,
    nowMs: Long,
    reminderConfigs: List<MachineReminderConfig>,
): Int {
    if (!machine.status.countsAsActiveInventory) return 0

    val configs = effectiveReminderConfigs(machine.id, reminderConfigs)
    if (configs.isEmpty()) return 0
    val lastTaskPlayCounts = lastTaskPlayCounts(events)
    var count = 0

    configs.filter { it.enabled }.forEach { config ->
        when (config.mode) {
            MachineReminderMode.manualOnly -> Unit
            MachineReminderMode.playBased -> {
                val intervalPlays = config.intervalPlays ?: 0
                if (intervalPlays <= 0) return@forEach
                val baseline = lastTaskPlayCounts[config.taskType] ?: 0
                if ((currentPlayCount - baseline) >= intervalPlays) count += 1
            }

            MachineReminderMode.dateBased -> {
                val intervalDays = config.intervalDays ?: 0
                if (intervalDays <= 0) return@forEach
                val lastAt = latestEventDate(config.taskType, events)
                if (lastAt == null) {
                    count += 1
                } else {
                    val dueAt = lastAt + intervalDays * 24L * 60L * 60L * 1000L
                    if (nowMs >= dueAt) count += 1
                }
            }
        }
    }
    return count
}

private fun currentPlayCount(eventsDesc: List<MachineEvent>): Int {
    val asc = eventsDesc.sortedWith(compareBy<MachineEvent> { it.occurredAtMs }.thenBy { it.createdAtMs }.thenBy { it.id })
    var runningTotal = 0
    asc.forEach { event ->
        event.loggedPlayCountTotal?.let { runningTotal = it }
    }
    return runningTotal
}

private fun lastTaskPlayCounts(eventsDesc: List<MachineEvent>): Map<MachineReminderTaskType, Int> {
    val asc = eventsDesc.sortedWith(compareBy<MachineEvent> { it.occurredAtMs }.thenBy { it.createdAtMs }.thenBy { it.id })
    var runningPlayCount = 0
    val lastByTask = mutableMapOf<MachineReminderTaskType, Int>()
    asc.forEach { event ->
        event.loggedPlayCountTotal?.let { runningPlayCount = it }
        MachineReminderTaskType.entries.forEach { taskType ->
            if (taskType.matchingEventTypes.contains(event.type)) {
                lastByTask[taskType] = runningPlayCount
            }
        }
    }
    return lastByTask
}

private fun latestEventDate(taskType: MachineReminderTaskType, events: List<MachineEvent>): Long? {
    return events.firstOrNull { taskType.matchingEventTypes.contains(it.type) }?.occurredAtMs
}

private fun effectiveReminderConfigs(
    machineID: String,
    reminderConfigs: List<MachineReminderConfig>,
): List<MachineReminderConfig> {
    val existing = reminderConfigs.filter { it.ownedMachineID == machineID }
    if (existing.isEmpty()) {
        return MachineReminderConfig.defaultConfigs(machineID).sortedBy { it.taskType.name }
    }

    val mergedByTask = MachineReminderConfig.defaultConfigs(machineID)
        .associateBy { it.taskType }
        .toMutableMap()
    existing.forEach { config ->
        mergedByTask[config.taskType] = config
    }
    return mergedByTask.values.sortedBy { it.taskType.name }
}
